text/3485-feature-documentation.md

```diff
@@ -0,0 +1,214 @@
+
+This RFC proposes adding feature documentation to `Cargo.toml`, which will allow
+for keeping feature definitions and documentation together.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A new `doc` key will be allowed within a feature's table. This key provides a
+markdown docstring describing the feature. The first paragraph will be treated
+as a summary, and should be suitable to display standalone without the rest of
+the description.
+
+Don't include the name of the feature in the summary; tools will typically
+already display this documentation alongside the name of the feature.
+
+```toml
+[features]
+# Feature without documentation
+foo = []
+
+# Short documentation comment
+bar = { enables = ["foo"], doc = "simple docstring here"}
+
+# Tables are preferred for longer descriptions
+[features.corge]
+enables = ["bar", "baz"]
+doc = """
+The first paragraph is a short summary, which might be displayed standalone.
+This could be a longer description of this feature
+"""
+```
+
+See [`feature-metadata`] for information about `enables`.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The new `doc` key accepts markdown-flavored text, and should be thought of as
+the equivalent to a `#[doc(...)]` attribute. Like doc comments, the first line
+should be treated as a summary. Intra-doc link support is not included in this
+RFC, so they should not be used.
+
+There is nothing in this RFC that cargo **must** do with this action, since it
+is mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it
+can be used for general diagnostic information such as during `cargo add` or a
+possible `cargo info` command. A sample application with `cargo add`:
+
+```text
+crab@rust foobar % cargo add regex
+    Updating crates.io index
+      Adding regex v1.7.3 to dependencies.
+             Features:
+             + perf             Enables all performance related features
+             + perf-dfa         Enables the use of a lazy DFA for matching
+             + perf-inline      Enables the use of aggressive inlining inside
+                                match routines
+             + perf-literal     Enables the use of literal optimizations for
+                                speeding up matches
+             + std              When enabled, this will cause regex to use the
+                                standard library
+             + unicode          Enables all Unicode features
+
+    Updating crates.io index
+```
+
+_(features like `aho-corasick`, `memchr`, or `use_std` would likely be
+`public = false` since they aren't listed on the crate landing page)_
+
+Any tools that want the information in `doc` will require access to the
+manifest. Adding this information to the index was decided against due to
+concerns about bloat, but this is further discussed in
+[future possibilities][future-possibilities].
```

Comment:
@Turbo87 why did you add crates.io team to this RFC?  Without Index support, this seems like the Cargo team owns the decision making on this while input for how crates.io may want to use this in their UX would still be welcome

Reply:
because we were already left out of the discussion in https://github.com/rust-lang/rfcs/pull/3416 and now we have to deal with a situation where old cargo won't be able to read newer crates on crates.io that use the new syntax. for now cargo could still normalize the new to the old syntax, but once we add additional fields (as proposed by this RFC), the normalization won't work anymore without losing that information when uploading it to crates.io.

IMHO it makes sense to include all of the relevant teams in RFCs such as this. if it turns out that our part is only small then it should be easy for our team to approve quickly.

Reply:
@Turbo87, perhaps there is a misunderstanding about how the RFC process is supposed to work, or what it means to have a team label on an RFC? The labels are part of the rfcbot system which will require approval for the RFC. That is intended for teams which have responsibility for the given content. In this case, this is a change to the cargo manifest which is the responsibility of the cargo team. I understand you would like to be kept in the loop for these kinds of changes, but that is the purpose of opening a public RFC, to allow everyone to see and comment on proposed changes. If you would like, we can let the crates.io team know on Zulip or tag the team on the RFC when there is an RFC that proposes a change to the manifest. Would that help?

Reply:
> the cargo manifest which is the responsibility of the cargo team

I think this is the main problem/misunderstanding here. I consider the crate file manifest to be part of the interface between cargo and crates.io and IMHO that means both teams are responsible, at least for the parts that are relevant to the registry.

I'm not sure an RFC is the right venue to discuss this though, since it's somewhat off-topic and distracting from the technical discussion, but I'm happy to participate in a responsibilities discussion if we can find a neutral moderator for it.

> If you would like, we can let the crates.io team know on Zulip or tag the team on the RFC when there is an RFC that proposes a change to the manifest.

that would definitely be appreciated :)